### PR TITLE
add bookmarks

### PR DESCRIPTION
**Pull request by davgur**

### CODE REVIEW

File: src/components/Sections/Personal/Bookmarks/Bookmarks/helper.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    const articleName = path.splice(-1);
+    return `${articleName} ${path.join('. ')}`;
+  }
+
+  if (content_type === CT_LIKUTIM) {
+    return name;
+  }
+
+  const collection = Object.values(collections)[0];
+  const part       = Number(collection?.ccuNames[cu.id]);
```
Comment (bbfsdev) [inline src/components/Sections/Personal/Bookmarks/Bookmarks/helper.js:27]: Are you sure this is a number here?
What happens when collection does not exist?

File: src/components/Sections/Personal/Bookmarks/Folders/Item.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+import { actions } from '../../../../../redux/modules/my';
+import { MY_NAMESPACE_FOLDERS } from '../../../../../helpers/consts';
+import { getMyItemKey } from '../../../../../helpers/my';
+import { stopBubbling } from '../../../../../helpers/utils';
+import { selectors as settings } from '../../../../../redux/modules/settings';
+import { getLanguageDirection } from '../../../../../helpers/i18n-utils';
+
+const FolderItem = ({ folder, selectedId, selectFolder, t }) => {
+  const [edit, setEdit]       = useState();
```
Comment (bbfsdev) [inline src/components/Sections/Personal/Bookmarks/Folders/Item.js:15]: useState should get default value?

File: public/locales/he/common.json [hunk 3]
```diff
@@ -777,6 +788,23 @@
...
+      "removeBookmark": "להסיר סימניה",
+      "searchBookmarks": "לחפש בסימניות",
+      "searchFolders": "חפש תיקיות",
+      "folders": "תיקיות",
+      "newFolder": "תיקיה חדשה",
+      "allFolders": "הכל",
+      "filterByFolder": "חפש ב",
+      "bookmarkUpdated": "הסימניה עודכנה בהצלחה!",
+      "bookmarkCreated": "הסימניה נוצרה בהצלחה!",
+      "confirmRemove": "האם ברצונך למחוק את {{name}}?",
```
[the remark this answers is not in the document]
Comment (bbfsdev) [reply public/locales/he/common.json:806]: You should duplicate this once for folder and once for bookmarks.
Or add another parameter {{what}}

File: src/components/Sections/Personal/Bookmarks/Bookmarks/Item.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+          <span className="separator">|</span>
+          <span className="source_name">
+            {title}
+          </span>
+        </List.Header>
+        <List.Description as={'em'}>
+          {citates.join(' ... ')}
+        </List.Description>
+        <List.Description>
+          {folders.map(renderFolder)}
```
Comment (bbfsdev) [inline src/components/Sections/Personal/Bookmarks/Bookmarks/Item.js:75]: image

- Too much space below folders. 
- Clicking on folders selects the bookmark and not the folder.

File: src/components/Sections/Personal/Bookmarks/Folders/List.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+import { actions, selectors } from '../../../../../redux/modules/my';
+import { MY_BOOKMARK_FILTER_FOLDER_ID, MY_NAMESPACE_FOLDERS } from '../../../../../helpers/consts';
+import { DeviceInfoContext } from '../../../../../helpers/app-contexts';
+import { actions as filtersActions, selectors as filtersSelectors } from '../../../../../redux/modules/bookmarkFilter';
+import FolderItem from './Item';
+import clsx from 'clsx';
+
+const FolderList = ({ t, close }) => {
+  const [editFolder, setEditFolder]         = useState(false);
+  const [query, setQuery]                   = useState();
```
Comment (bbfsdev) [inline src/components/Sections/Personal/Bookmarks/Folders/List.js:15]: Again empty useState, is this a common practice?

File: src/components/Sections/Personal/Bookmarks/Folders/List.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+  const { isMobileDevice } = useContext(DeviceInfoContext);
+
+  const items = useSelector(state => selectors.getList(state.my, MY_NAMESPACE_FOLDERS)).filter(x => !query || x.name.toLowerCase().includes(query));
+
+  const dispatch = useDispatch();
+  useEffect(() => {
+    if (items.length === 0)
+      dispatch(actions.fetch(MY_NAMESPACE_FOLDERS, { 'order_by': 'id DESC' }));
+    // eslint-disable-next-line  react-hooks/exhaustive-deps
```
Comment (bbfsdev) [inline src/components/Sections/Personal/Bookmarks/Folders/List.js:26]: Why disabling exhaustive deps is okay here?

File: src/components/shared/SaveBookmark/BookmarkForm.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+  const bookmark = useSelector(state => selectors.getItemByKey(state.my, MY_NAMESPACE_BOOKMARKS, key));
+  const items = useSelector(state => selectors.getList(state.my, MY_NAMESPACE_FOLDERS)).filter(x => !query || x.name.toLowerCase().includes(query));
+  const saved = items.filter(f => bookmark?.folder_ids?.includes(f.id)).map(f => f.id);
+
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (items.length === 0)
+      dispatch(actions.fetch(MY_NAMESPACE_FOLDERS, { 'order_by': 'id DESC' }));
+    // eslint-disable-next-line  react-hooks/exhaustive-deps
```
Comment (bbfsdev) [inline src/components/shared/SaveBookmark/BookmarkForm.js:40]: Why is this okay and in all other places.

File: src/components/shared/SaveBookmark/BookmarkForm.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  }, [dispatch]);
+
+  useEffect(() => {
+    if (selected === null) {
+      setSelected([...saved]);
+    }
+    // eslint-disable-next-line  react-hooks/exhaustive-deps
```
Comment (bbfsdev) [inline src/components/shared/SaveBookmark/BookmarkForm.js:47]: Why disable lint is okay?

File: src/redux/modules/mdb.js [hunk 9]
```diff
@@ -572,6 +572,8 @@ const getDenormContentUnit = (state, id) => {
   return cu;
 };
 
+const nestedGetDenormContentUnit = state => id => getDenormContentUnit(state, id);
```
Comment (bbfsdev) [inline src/redux/modules/mdb.js:575]: Why do you need this?

File: src/components/Pages/Unit/widgets/UnitMaterials/UnitBar.js [hunk 10]
```diff
@@ -3,13 +3,17 @@ import { Button } from 'semantic-ui-react';
...
+import BookmarkButton from '../../../../shared/SaveBookmark/BookmarkButton';
 
-const UnitBar = ({ fontSize = 0, handleSettings }) => {
+const UnitBar = ({ fontSize = 0, handleSettings, source }) => {
 
   const print = () => window.print();
 
   return (
     <div className="source__header-toolbar">
+      {
```
Comment (bbfsdev) [inline src/components/Pages/Unit/widgets/UnitMaterials/UnitBar.js:14]: image

In mobile icon overflows title of source.

File: src/components/Sections/Personal/Bookmarks/HeaderMobile.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            basic
+            icon
+            className="no-border"
+          >
+            {folder.name}
+            <Button
+              basic
+              compact
+              className="no-padding no-shadow"
+              onClick={removeFolderFilter}
```
Comment (bbfsdev) [inline src/components/Sections/Personal/Bookmarks/HeaderMobile.js:73]: image

Make the button larger. Consider putting  the folder name in the same row and ellipsis the text

File: src/components/shared/DocToolbar/Toolbar.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+                <Header.Content>
+                  {t('share-text.docbar-title')}
+                </Header.Content>
+              </Header>
+              <Menu inverted borderless>
+                <ShareBtn url={url}/>
+                <CopyLinkBtn text={url}/>
+                <CopyTextBtn text={text}/>
+                {source && <BookmarkBtn source={source} close={handleToggle}/>}
+                {/*<TagBtn />*/}
```
Comment (bbfsdev) [inline src/components/shared/DocToolbar/Toolbar.js:63]: What are those two? Do we need them?

**Review by bbfsdev — CHANGES_REQUESTED** [review]
image

Still many lint errors. Please check.

image

Save bookmark on mobile is too wide... 

Is there an option to select and add "custom text" bookmark on mobile? I was not able to select some text and get the bookmark widget.

File: src/components/shared/SaveBookmark/BookmarkForm.js (hunk 8 above)
Comment (davgur) [reply src/components/shared/SaveBookmark/BookmarkForm.js:47]: //update selected only when saved was changed (that mean only on server response)
    //and not on changes of selected

add comment

File: src/components/shared/DocToolbar/Toolbar.js (hunk 12 above)
Comment (davgur) [reply src/components/shared/DocToolbar/Toolbar.js:63]: temporary hidden will used on next PR

remove here

File: src/components/Pages/Unit/widgets/UnitMaterials/UnitBar.js (hunk 10 above)
Comment (davgur) [reply src/components/Pages/Unit/widgets/UnitMaterials/UnitBar.js:14]: its separated bug

File: src/redux/modules/mdb.js (hunk 9 above)
Comment (davgur) [reply src/redux/modules/mdb.js:575]: I sue it [here](https://github.com/Bnei-Baruch/kmedia-mdb/blob/d7b66b566d155f2f397982d13e74ddd8df153318/src/components/Sections/Personal/Bookmarks/Bookmarks/List.js#L44) 
For reuse, and can use on _if_ denorm function with single param _id_

File: public/locales/he/common.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (davgur) [reply public/locales/he/common.json:806]: separate translations

File: src/components/Sections/Likutim/Likut.js [hunk 13]
```diff
@@ -20,7 +20,7 @@ import MenuLanguageSelector from '../../../components/Language/Selector/MenuLang
 import Link from '../../../components/Language/MultiLanguageLink';
 import WipErr from '../../shared/WipErr/WipErr';
 import Download from '../../shared/Download/Download';
-import ScrollToSearch from '../../shared/ScrollToSearch';
+import ScrollToSearch from '../../shared/DocToolbar/ScrollToSearch';
 
 // expected unit of type Likutim
 const Likut = ({ t }) => {
```
[the remark this answers is not in the document]
Comment (davgur) [reply src/components/Sections/Likutim/Likut.js:26]: put minimaze to the other place

File: src/components/shared/SaveBookmark/BookmarkForm.js [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+      <Checkbox
+        checked={selected?.includes(f.id)}
+        onChange={(e, { checked }) => handleChange(checked, f.id)}
+        label={f.name}
+      />
+    </List.Item>
+  );
+
+  return (
+    <>
```
[the remark this answers is not in the document]
Comment (davgur) [reply src/components/shared/SaveBookmark/BookmarkForm.js:119]: you mean <> === <React.Fragment>?

File: src/components/shared/SubscribeBtn.js [hunk 15]
```diff
@@ -49,6 +49,7 @@ const SubscribeBtn = ({ unit = {}, t, collection }) => {
     if (!sub && (subsByType || subsByCO)) {
       dispatch(actions.fetch(MY_NAMESPACE_SUBSCRIPTIONS, { addToList: false, ...subParams }));
     }
+    // eslint-disable-next-line  react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (davgur) [reply src/components/shared/SubscribeBtn.js:52]: subsByType and subsByCO, subParams are depended by key, so was remove additional dependency

File: public/locales/cs/common.json [hunk 16]
```diff
@@ -690,6 +701,24 @@
...
+      "removeBookmark": "Remove Bookmark",
+      "searchBookmarks": "Search Bookmarks",
+      "searchFolders": "Search Folders",
+      "folders": "Folders",
+      "newFolder": "New Folder",
+      "allFolders": "All",
+      "filterByFolder": "Search in",
+      "bookmarkUpdated": "Bookmark updated successfully!",
+      "bookmarkCreated": "Bookmark successfully created!",
+      "confirmRemoveBookmark": "Are you sure you want to delete bookmark {{name}}?",
```
Comment (bbfsdev) [inline public/locales/cs/common.json:719]: Add "" around the name.

File: src/components/Sections/Personal/Bookmarks/Folders/Item.js (hunk 2 above)
Comment (bbfsdev) [reply src/components/Sections/Personal/Bookmarks/Folders/Item.js:15]: ?

File: src/components/shared/SaveBookmark/BookmarkForm.js (hunk 8 above)
Comment (bbfsdev) [reply src/components/shared/SaveBookmark/BookmarkForm.js:48]: So why not add all the parameters and create the correct condition (if) on line 43?

File: src/components/shared/SaveBookmark/BookmarkForm.js (hunk 7 above)
Comment (bbfsdev) [reply src/components/shared/SaveBookmark/BookmarkForm.js:39]: ?

**Review by bbfsdev — CHANGES_REQUESTED** [review]
https://staging-archive.kabbalahmedia.info/he/sources/4sU63POC?language=ru&srchend=%D0%B2%D0%BE%D0%B7%D0%BD%D0%B0%D0%B3%D1%80%D0%B0%D0%B6%D0%B4%D0%B5%D0%BD%D0%B8%D0%B5%20%D1%81%D0%BE%D1%81%D1%82%D0%BE%D0%B8%D1%82%20%D0%B2%20%D1%82%D0%BE%D0%BC%2C%20%D1%87%D1%82%D0%BE%3A%24%3A420&srchstart=%D1%81%D1%82%D1%80%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D1%8F%20%D0%B4%D0%BE%D1%80%D0%BE%D0%B3%D0%B8%20%D1%82%D0%B5%D0%B1%D0%B5%3F%C2%BB%20%D0%9E%D1%82%D0%B2%D0%B5%D1%82%D0%B8%D0%BB%20%D0%BE%D0%BD%3A%24%3A66

This link does not highlight the relevant text

**Review by bbfsdev — APPROVED** [review]

**Pull request merged**